ibis/expr/operations.py

```diff
@@ -12,25 +12,228 @@
+        expression._root_tables() for expression in expressions
+    )
+    return list(toolz.unique(roots, key=id))
+
+
+class TypeSignature(OrderedDict):
+
+    __slots__ = tuple()
+
+    def __call__(self, *args, **kwargs):
+        if len(args) > len(self.keys()):
+            raise TypeError('takes {} positional arguments but {} were '
+                            'given'.format(len(args), len(self.keys())))
+
+        result = []
+        for i, (name, rule) in enumerate(self.items()):
+            if i < len(args):
+                if name in kwargs:
+                    raise TypeError('got multiple values for argument'
+                                    '{}'.format(name))
+                value = rule.call(args[i])
+            elif name in kwargs:
+                value = rule.call(kwargs[name])
+            else:
+                value = rule.call()
+
+            result.append((name, value))
+
+        return result
+
+    def names(self):
+        return self.keys()
+
+
+class OperationMeta(type):
+
+    @classmethod
+    def __prepare__(metacls, name, bases, **kwds):
+        return OrderedDict()
+
+    def __new__(cls, name, parents, dct):
+        signature = TypeSignature()
+
+        # inherit from parent signatures
+        for parent in parents:
+            if hasattr(parent, 'signature'):
+                signature.update(parent.signature)
+
+        attrs = OrderedDict()
+        for k, v in dct.items():
+            if isinstance(v, rlz.validator):
+                signature[k] = v
+            elif isinstance(v, type):
+                signature[k] = rlz.instanceof(v)
+            else:
+                attrs[k] = v
+
+        attrs['signature'] = signature
+        attrs['__slots__'] = tuple(signature.keys()) + ('_expr_cached',)
+
+        return super(OperationMeta, cls).__new__(cls, name, parents, attrs)
+
+
+class Node(six.with_metaclass(OperationMeta, object)):
+
+    def __init__(self, *args, **kwargs):
+        self._expr_cached = None
+        for name, value in self.signature(*args, **kwargs):
+            setattr(self, name, value)
+        self._validate()
+
+    def _validate(self):
+        pass
+
+    def __repr__(self):
+        return self._repr()
+
+    def _repr(self, memo=None):
+        if memo is None:
+            from ibis.expr.format import FormatMemo
+            memo = FormatMemo()
+
+        opname = type(self).__name__
+        pprint_args = []
+
+        def _pp(x):
+            return _safe_repr(x, memo=memo)
+
+        for x in self.args:
+            if isinstance(x, (tuple, list)):
+                pp = repr([_pp(y) for y in x])
+            else:
+                pp = _pp(x)
+            pprint_args.append(pp)
+
+        return '%s(%s)' % (opname, ', '.join(pprint_args))
+
+    def blocks(self):
+        # The contents of this node at referentially distinct and may not be
+        # analyzed deeper
+        return False
+
+    def flat_args(self):
+        for arg in self.args:
```

Comment:
I feel like we have a `flatten` function laying around somewhere.

Reply:
There [isn't](https://github.com/ibis-project/ibis/search?utf8=%E2%9C%93&q=flatten&type=) one, should it go to `ibis.util`?

Reply:
Let's leave this here for now.

Reply:
One wrinkle here is that this only flattens one nesting level. So a general `flatten` function would probably break some things. Let's revisit `flatten` after this PR is merged.